### PR TITLE
Use ids instead of slugs

### PR DESCRIPTION
**Pull request by Tristramg**
Problem : slugs change a lot, often resulting in broken dataset imports.

Solution : use ids instead of slugs.

### CODE REVIEW

File: lib/mix/tasks/transport/validate_data.ex [hunk 1]
```diff
@@ -19,16 +19,16 @@ defmodule Mix.Tasks.Transport.ValidateData do
...
     |> Enum.each(fn(dataset) ->
-      %{project: project, name: dataset.slug, url: dataset.download_uri}
+      %{project: project, name: dataset.id, url: dataset.download_uri}
       |> DataValidation.create_feed_source
       |> case do
         {:ok, feed_source} ->
           :ok = DataValidation.validate_feed_source(%{project: project, feed_source: feed_source})
           Logger.info(" <message>  Validating dataset")
-          Logger.info(" <slug>     #{dataset.slug}")
+          Logger.info(" <id>     #{dataset.id}")
```
Comment (maukoquiroga) [inline lib/mix/tasks/transport/validate_data.ex]: For logging purposes, do we want the `id` or the `slug` ?

File: lib/mix/tasks/transport/validate_data.ex [hunk 2]
```diff
@@ -19,16 +19,16 @@ defmodule Mix.Tasks.Transport.ValidateData do
...
         {:error, error} ->
           Logger.error("<message>  Unable to create feed source: #{inspect(error)}")
-          Logger.error("<slug>     #{dataset.slug}")
+          Logger.error("<id>     #{dataset.id}")
```
Comment (maukoquiroga) [inline lib/mix/tasks/transport/validate_data.ex]: Idem supra.

File: lib/transport/import_data_service.ex [hunk 3]
```diff
@@ -8,25 +8,25 @@ defmodule Transport.ImportDataService do
...
 
-  def import_from_udata(slug) do
+  def import_from_udata(id) do
     base_url = Application.get_env(:transport, :datagouvfr_site)
-    url      = "#{base_url}/api/1/datasets/#{slug}/"
+    url      = "#{base_url}/api/1/datasets/#{id}/"
 
     Logger.info(" <message>  Importing dataset")
-    Logger.info(" <slug>     #{slug}")
+    Logger.info(" <id>       #{id}")
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: Idem supra.

File: lib/transport/import_data_service.ex [hunk 4]
```diff
@@ -36,7 +36,7 @@ defmodule Transport.ImportDataService do
     else
       {:error, error} ->
         Logger.error("<message>  #{inspect error}")
-        Logger.error("<slug>     #{slug}")
+        Logger.error("<id>       #{id}")
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: Idem supra.

File: lib/transport/import_data_service.ex [hunk 5]
```diff
@@ -48,7 +48,7 @@ defmodule Transport.ImportDataService do
   def get_dataset(%{} = dataset) do
     dataset =
       dataset
-      |> Map.take(["title", "description", "license", "slug"])
+      |> Map.take(["title", "description", "license", "id"])
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: We want them both no?

File: lib/transport/reusable_data/dataset.ex [hunk 6]
```diff
@@ -4,15 +4,15 @@ defmodule Transport.ReusableData.Dataset do
   reuser.
   """
 
-  defstruct [
+defstruct [
```
Comment (maukoquiroga) [inline lib/transport/reusable_data/dataset.ex]: > defstruct [

  defstruct [

File: lib/transport/reusable_data/dataset.ex [hunk 7]
```diff
@@ -4,15 +4,15 @@ defmodule Transport.ReusableData.Dataset do
...
     :_id,
     :title,
     :description,
     :logo,
     :spatial,
     :coordinates,
     :licence,
-    :slug,
+    :id,
```
Comment (maukoquiroga) [inline lib/transport/reusable_data/dataset.ex]: We want them both no?

File: lib/transport/reusable_data/reusable_data.ex [hunk 8]
```diff
@@ -164,7 +164,7 @@ defmodule Transport.ReusableData do
 
   def get_dataset_id(conn, dataset) do
```
Comment (maukoquiroga) [inline lib/transport/reusable_data/reusable_data.ex]: Why would be continue to use this function? I think it is a leftover.

Comment (maukoquiroga) [inline lib/transport/reusable_data/reusable_data.ex]: Why would be continue to use this function? I think it is a leftover.

File: lib/transport_web/api/controllers/dataset_controller.ex [hunk 9]
```diff
@@ -6,11 +6,11 @@ defmodule TransportWeb.API.DatasetController do
     render(conn, data: ReusableData.list_datasets)
   end
 
-  def show(%Plug.Conn{} = conn, %{"slug" => slug}) do
-    render(conn, data: ReusableData.get_dataset(slug))
+  def show(%Plug.Conn{} = conn, %{"id" => id}) do
```
Comment (maukoquiroga) [inline lib/transport_web/api/controllers/dataset_controller.ex]: We want to use `id` while communicating internally with data.gouv.fr. But do we want to do it in our front-end ?

File: lib/transport_web/api/serializers/dataset_serializer.ex [hunk 10]
```diff
@@ -4,13 +4,8 @@ defmodule TransportWeb.API.DatasetSerializer do
   """
 
   use TransportWeb, :serializer
-  alias BSON.ObjectId
-  alias Transport.ReusableData.Dataset
 
-  location "/datasets/:slug/"
-  attributes [:title, :slug, :coordinates, :validations]
+  location "/datasets/:id/"
```
Comment (maukoquiroga) [inline lib/transport_web/api/serializers/dataset_serializer.ex]: Same, this is our front's representation, it not the same context than our internal relationship with data.gouv.fr.

File: lib/transport_web/controllers/dataset_controller.ex [hunk 11]
```diff
@@ -11,8 +11,8 @@ defmodule TransportWeb.DatasetController do
     |> render("index.html")
   end
 
-  def new(%Plug.Conn{} = conn, %{"slug" => slug}) do
-    case Organizations.get(conn, slug) do
+  def new(%Plug.Conn{} = conn, %{"id" => id}) do
```
Comment (maukoquiroga) [inline lib/transport_web/controllers/dataset_controller.ex]: Same, this is our front's representation, it not the same context than our internal relationship with data.gouv.fr.

File: lib/transport_web/controllers/user_controller.ex [hunk 12]
```diff
@@ -40,7 +40,7 @@ defmodule TransportWeb.UserController do
       {:ok, response} ->
         conn
         |> put_flash(:info, dgettext("user", "Organization added"))
-        |> redirect(to: user_path(conn, :organization_datasets, response["slug"]))
+        |> redirect(to: user_path(conn, :organization_datasets, response["id"]))
```
Comment (maukoquiroga) [inline lib/transport_web/controllers/user_controller.ex]: Same, this is our front's representation, it not the same context than our internal relationship with data.gouv.fr.

File: lib/transport_web/router.ex [hunk 13]
```diff
@@ -41,19 +41,19 @@ defmodule TransportWeb.Router do
 
     scope "/datasets" do
       get "/", DatasetController, :index
-      get "/:slug/", DatasetController, :details
+      get "/:id/", DatasetController, :details
```
Comment (maukoquiroga) [inline lib/transport_web/router.ex]: Same, this is our front's representation, it not the same context than our internal relationship with data.gouv.fr.

File: lib/transport_web/templates/dataset/details.html.eex [hunk 14]
```diff
@@ -11,7 +11,7 @@
           <div class="shortlist__link shortlist__link--external-link">
             <i class="icon icon--link" aria-hidden="true"></i>
             <%= link(dgettext("page-shortlist", "See on data.gouv.fr"),
-                     to: "https://www.data.gouv.fr/datasets/#{@dataset.slug}/",
+                     to: "https://www.data.gouv.fr/datasets/#{@dataset.id}/",
```
Comment (maukoquiroga) [inline lib/transport_web/templates/dataset/details.html.eex]: Same, this is our front's representation, it not the same context than our internal relationship with data.gouv.fr.

File: test/transport_web/solutions/data_reuse/using_list_test.exs [hunk 15]
```diff
@@ -23,7 +23,7 @@ defmodule TransportWeb.Solution.DataReuse.UsingListTest do
       title: "Horaires et arrêts du réseau IRIGO - format GTFS",
       anomalies: [],
       coordinates: [-0.5630548425091684,47.47654241641714],
-      slug: "horaires-et-arrets-du-reseau-irigo-format-gtfs",
+      id: "horaires-et-arrets-du-reseau-irigo-format-gtfs",
```
Comment (maukoquiroga) [inline test/transport_web/solutions/data_reuse/using_list_test.exs]: This should fail...

File: test/transport_web/solutions/data_reuse/using_map_test.exs [hunk 16]
```diff
@@ -23,7 +23,7 @@ defmodule TransportWeb.Solution.DataReuse.UsingMapTest do
       title: "Horaires et arrêts du réseau IRIGO - format GTFS",
       anomalies: [],
       coordinates: [-0.5630548425091684,47.47654241641714],
-      slug: "horaires-et-arrets-du-reseau-irigo-format-gtfs",
+      id: "horaires-et-arrets-du-reseau-irigo-format-gtfs",
```
Comment (maukoquiroga) [inline test/transport_web/solutions/data_reuse/using_map_test.exs]: As well...

File: lib/mix/tasks/transport/validate_data.ex (hunk 1 above)
Comment (Tristramg) [reply lib/mix/tasks/transport/validate_data.ex]: BOTH! ;) I’m changing it for all

File: lib/transport_web/api/controllers/dataset_controller.ex (hunk 9 above)
Comment (Tristramg) [reply lib/transport_web/api/controllers/dataset_controller.ex]: I’d be in favor to use ids everywhere, excepted in what is shown to the user. That would be only the url.

Hence, here, I would keep the id, but adapt it bellow

**Comment by Tristramg** [comment]
All right! Thank you for your comments @maukoquiroga !

Now the id is only used to import the data. Otherwise we still use slugs.

**Pull request merged**